dfu/mcuboot: Add missing include

kurddt opened the pull request:
dfu/mcuboot.h uses some definitions of bootutil/bootutil_public.h but does not include the header.
fix #52095

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

nordicjm commented on modules/Kconfig.mcuboot_bootutil:
```diff
@@ -29,4 +29,12 @@ config BOOT_IMAGE_ACCESS_HOOKS
 	  It is up to the application project to add source file which
 	  implements hooks to the build.
 
+config APP_LINK_WITH_MCUBOOT_BOOTUTIL
+	bool "Link 'app' with MCUBOOT BOOTUTIL"
```
What is the purpose of this? It is not used

nandojve commented on subsys/mgmt/updatehub/CMakeLists.txt:
```diff
@@ -15,3 +15,5 @@ zephyr_library_sources_ifdef(CONFIG_UPDATEHUB_SHELL shell.c)
   ${ZEPHYR_BASE}/subsys/mgmt/updatehub/include
 )
+
+zephyr_library_link_libraries_ifdef(CONFIG_UPDATEHUB subsys__dfu__boot)
```
This seems little odd to me since we depends on MCUBOOT, see https://github.com/zephyrproject-rtos/zephyr/blob/67e8d03280add828d9e44801733efde236332025/subsys/mgmt/updatehub/Kconfig#L10

Question: Why that dependency not address this?

kurddt replied:
I actually don't know

kurddt replied:
I think I understand, the issue here is that dfu_boot depends on MCUBOOT_UTIL which unlike most of the zephyr libraries does not add it's public headers to the `zephyr_ interface` target by using `zephyr_include_directories`. 

If I understand correctly this target is added as dependency to all zephyr_libraries. That so that all include directories of all zephyr_libraries are added to the include path of all libraries when compiling. When it's not done (like for MCUBOOT_UTIL) the only way for the toolchain to know this dependency is by using "standard" cmake mechanism

Because of that the dependency on subsys__dfu__boot need to be made explicit.

I guess an other solution would be to use `zephyr_include_directories` instead of `target_include_directories` here: https://github.com/zephyrproject-rtos/mcuboot/blob/cfec947e0f8be686d02c73104a3b1ad0b5dcf1e6/boot/bootutil/zephyr/CMakeLists.txt#L9. But I don't know if this could lead to other issues

@nvlsianpu What is your opinion on that ?

carlescufi replied:
@de-nordic ^^

de-nordic replied:
This is external library so I do not think that its include paths should be part of Zephyr interface.
IMHO, problem with mcuboot.h is that it provides Zephyr subsystem API, but the API interface itself requires inclusion of external library.
Neither of `BOOT_SWAP_TYPE_` definitions are usable with API provided by the mcuboot.h as as there is no function or structure, or type to use them with.
The only somehow usable definition is `BOOT_IMG_VER_STRLEN_MAX`; somehow because it has nothing to do with MCUboot, as MCUboot version is bunch of integers. Everywhere where `BOOT_IMG_VER_STRLEN_MAX` is used to allocate buffer, custom function is used to fill the buffer with version: `hawkbit_get_firmware_version` or `updatehub_get_firmware_version`. So basically this define just limits how other subsystems format MCUboot version, but MCUboot does not use the format, nor actually define it. It can stay.

Problem is with `BOOT_TRAILER_IMG_STATUS_OFFS` which depends on bootutil_public.h for the definitions.
And this is the one that causes problems.

What I would suggest doing is:
1) move all `BOOT_SWAP_TYPE_` definitions into dfu internal mcuboot_priv.h file, that would only be included by dfu subsystem source;
2) include the bootutil_public.h on top of mcuboot_priv.h so that all the checks could still be done there (although I thing that C would detect incompatible redefinitions anyway);
3) The tricky part: how fix the main problem the `BOOT_TRAILER_IMG_STATUS_OFFS`?

For the 3) options are to redefine `BOOT_MAGIC_SZ` and `BOOT_MAX_ALIGN` it under these or other identifiers and add proper compatibility checks, for them, in mcuboot_priv.h, and at the same time avoid including bootutil_public.h here.
Or to stop pretending that `BOOT_TRAILER_IMG_STATUS_OFFS` does not depend on Flash Map API and define some function like
size_t `boot_get_trailer_status_offset(const struct flash_area *fa)` that will be doing the same job, but will not require inclusion of `bootutil_public.h` header at the level of mcuboot.h.

kurddt replied:
> Neither of BOOT_SWAP_TYPE_ definitions are usable with API provided by the mcuboot.h as as there is no function or structure, or type to use them with.

I don't entirely understand this point: `BOOT_SWAP_TYPE_` are returned by `mcuboot_swap_type_multi` and `mcuboot_swap_type` which are used outside of the dfu subsystem: https://github.com/zephyrproject-rtos/zephyr/blob/a0380aabff80c2476c7cdd6bc0844310ce4298ce/subsys/mgmt/mcumgr/grp/img_mgmt/src/zephyr_img_mgmt.c#L496

So unless those two function are removed from the API it's not possible to move their definition to a private include.

As a side note I also use those definitions upstream to implement an alternative to the smp server for the image management that better fit my needs

de-nordic replied:
> `mcuboot_swap_type_multi`

Sorry, I missed that one.

OK then so we reduce mcuboot.h to:
```
/** Attempt to boot the contents of slot 0. */
#define BOOT_SWAP_TYPE_NONE     1
/** Swap to slot 1.  Absent a confirm command, revert back on next boot. */
#define BOOT_SWAP_TYPE_TEST     2
/** Swap to slot 1, and permanently switch to booting its contents. */
#define BOOT_SWAP_TYPE_PERM     3
/** Swap back to alternate slot.  A confirm changes this state to NONE. */
#define BOOT_SWAP_TYPE_REVERT   4
/** Swap failed because image to be run is not valid */
#define BOOT_SWAP_TYPE_FAIL     5

#define BOOT_IMG_VER_STRLEN_MAX 25  /* 255.255.65535.4294967295\0 */

#define BOOT_MAGIC_SZ <num>
#define BOOT_MAX_ALIGN <num>

#define BOOT_TRAILER_IMG_STATUS_OFFS(bank_area) ((bank_area)->fa_size -\
                                                  BOOT_MAGIC_SZ -\
                                                  BOOT_MAX_ALIGN * 2)

```

At this point you do not even have to redefine these in mcuboot_priv.h. Just use the fact that different definitions with the same identifier will cause collision and include the mcuboot.h and the bootutil_public.h in one file, where it will be checked automatically whether there are not redefinitions; happens here already:
https://github.com/zephyrproject-rtos/zephyr/blob/5d059117fd279ee0eec4fb1c4d0244a37937013c/subsys/dfu/boot/mcuboot.c#L19-L20

MCUBOOT_BOOTUTIL should not be required to be linked by app as long as app does not directly use its services. DFU uses MCUBOOT_UTIL and provides services available through mcuboot.h, but DFU does link the library here:
https://github.com/zephyrproject-rtos/zephyr/blob/5d059117fd279ee0eec4fb1c4d0244a37937013c/subsys/dfu/boot/CMakeLists.txt#L7

kurddt replied:
Ok so if I follow your suggestion for 3) we can either: 
* Revert https://github.com/zephyrproject-rtos/zephyr/pull/50371
* Change the BOOT_TRAILER_IMG_STATUS_OFFS macro to a function

Personally I would be in favor to change the macro to a function

de-nordic replied:
Then lets make it a function. It will require flash_map.h for the flash_area definition, but this is Zephyr provided subsystem and we do not have to link anything extra.

This will also allow us to drop MCUBOOT_BOOTUTIL here https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/dfu/img_util/CMakeLists.txt.

kurddt replied:
@de-nordic To keep the API more uniform. The function could take and `area_id` instead of a `flash_area` structure

kurddt replied:
I agree with you reasoning. But that will make almost mandatory for a downstream user (this API is public) to open a flash area before end, while the other function do not. 

I don't think that the re-opening of the flash_area will have much performance impact on the current user of this function in the zephyr tree. But I can also add the two "flavors" of the function one that take an `area_id` and one the size

de-nordic replied:
> I agree with you reasoning. But that will make almost mandatory for a downstream user (this API is public) to open a flash area before end, while the other function do not.
> 

Same as the macro did, which required flash area to be opened to have flash_area object pointer.
You are able to to use `FIXED_PARTITION_SIZE(<dts_node_label>)` macro to obtain size of partition, this is the same `<dts_node_label>` that is used with `FIXED_PARTITION_ID`, given to `flash_area_open`. Of course when you need to iterate over identifiers you have stored somewhere.

> I don't think that the re-opening of the flash_area will have much performance impact on the current user of this function in the zephyr tree. But I can also add the two "flavors" of the function one that take an `area_id` and one the size

If you do not mind, please provide the two "flavors".  There is a little bit of performance hit because `flash_area_open` iterates over array of partitions to find the area; this isn't huge but is there anyway. Also in case when we make `flash_area_open` open for exclusive access, second call on flash area may fail.

Commit:
mcuboot: Remove public dependency on bootutil

BOOT_MAGIC_SZ and BOOT_MAX_ALIGN were used in the header without
including bootutil/bootutil_public.h. This change remove the need of
the inclusion by making the dependency private.
Fixes #52095

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

de-nordic commented on include/zephyr/dfu/mcuboot.h:
```diff
@@ -248,6 +247,22 @@ int boot_request_upgrade_multi(int image_index, int permanent);
+ * @brief Get the offset of the status from an image bank size
+ *
+ * @param area_size size of image bank
+ * @return offset of the status
```
Add note that negative value means that the trailer will not fit into given size and user needs to take care so that data of an image will not cross the offset.

de-nordic commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -242,3 +241,29 @@ int boot_erase_img_bank(uint8_t area_id)
+		return rc;
+	}
+
+	offset = boot_get_trailer_status_offset(fa->fa_size);
+	if (offset < 0) {
+		return -EFAULT
+	}
```
`flash_area_close` missing on error.

de-nordic commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -242,3 +241,29 @@ int boot_erase_img_bank(uint8_t area_id)
+	int rc;
+	const struct flash_area *fa;
+	ssize_t offset;
+
```
One empty line is enough.

de-nordic reviewed (changes requested):
Minor comments.
Please add additional commit for the release notes where removal of `BOOT_TRAILER_IMG_STATUS_OFFS` with new replacements.

Add: Fixes https://github.com/zephyrproject-rtos/zephyr/issues/52095
line to original commit message (this should automatically close the issue when merged).

kurddt commented:
Could you indicate where in the changelog you want to have this change documented? Should it be under MCUBoot (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.3.rst#mcuboot) or removed API (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.3.rst#removed-apis-in-this-release)

de-nordic commented:
> Could you indicate where in the changelog you want to have this change documented? Should it be under MCUBoot (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.3.rst#mcuboot) or removed API (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.3.rst#removed-apis-in-this-release)

DFU section of drivers, it seems that previous changes to this file have been reported there.

Commit:
doc: release-notes: Add note on mcuboot dfu changes

Add a note on removed BOOT_TRAILER_IMG_STATUS_OFFS macro and its
replacement functions

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

nordicjm commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -251,3 +250,30 @@ int boot_erase_img_bank(uint8_t area_id)
+
+ssize_t boot_get_trailer_status_offset(size_t area_size)
+{
+	return area_size - BOOT_MAGIC_SZ - BOOT_MAX_ALIGN * 2;
```
Missing cast on area_size?

nordicjm reviewed (commented):
Typo in commit message *This change remove the need

And this is stable API change so https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes needs to be followed

kurddt commented:
> Typo in commit message *This change remove the need
> 
> And this is stable API change so https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes needs to be followed

@nordicjm To make it backward compatible, I can change the macro so that it calls the function. But I will open a RFC PR

de-nordic commented:
> Typo in commit message *This change remove the need
> 
> And this is stable API change so https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes needs to be followed

I do not think so. I think that this file falls under "gray area" because even though the mcuboot.h is public API, due to location at zephyr/include, it has been missing tags an has never been documented.

nordicjm reviewed (commented):
typo in commit message *were

stephanosio commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -239,6 +239,7 @@ Drivers and Sensors
 * DAC
 
 * DFU
+  * Remove :c:macro:`BOOT_TRAILER_IMG_STATUS_OFFS` in favor a two new functions; :c:func:`boot_get_area_trailer_status_offset` and :c:func:`boot_get_trailer_status_offset`
```
Please keep the line length to under 100.

de-nordic commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -9,7 +9,6 @@
 #include <errno.h>
 #include <string.h>
 #include <zephyr/drivers/flash.h>
-#include <zephyr/storage/flash_map.h>
```
Yes, it s.
I a not sure whether #include <zephyr/drivers/flash.h> is as there is no direct flash access.
